### ISSUE TITLE
th-desugar doesn't desugar partial pattern-matches in do-notation to `fail`

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#80]
Take this function:

```haskell
f :: Maybe Int
f = do
  Just x <- Nothing
  return x
```

Despite its appearance, `f` is a total function, since the apparently partial pattern-match desugars to something resembling:

```haskell
case Nothing of
  Just x -> return x
  Nothing -> fail "Pattern match failure in do expression"
```

However, `th-desugar` doesn't respect this property during its own desugaring, as demonstrated in this GHCi session:

```
λ> putStrLn $([d| f :: Maybe Int; f = do { Just x <- Nothing; return x } |] >>= dsDecs >>= stringE . groom)
[DLetDec
   (DSigD f_6989586621679077736
      (DAppT (DConT GHC.Base.Maybe) (DConT GHC.Types.Int))),
 DLetDec
   (DValD (DVarPa f_6989586621679077736)
      (DAppE (DAppE (DVarE GHC.Base.>>=) (DConE GHC.Base.Nothing))
         (DLamE [arg_6989586621679077738_6989586621679077739]
            (DCaseE (DVarE arg_6989586621679077738_6989586621679077739)
               [DMatch (DConPa GHC.Base.Just [DVarPa x_6989586621679077737])
                  (DAppE (DVarE GHC.Base.return) (DVarE x_6989586621679077737))]))))]
```

If you squint at that long enough, you'll eventually realize that it's desugaring `Just x <- Nothing` to `case Nothing of { Just x -> return x }`, which is less defined than it should be (after all, the overridden `fail` implementation in the `Monad []` instance is what makes this trick work). We should implement the `fail` desugaring in `th-desugar` as well to compensate for this.

**Comment by RyanGlScott** [comment]
Note that this issue [blocks `singletons`](https://github.com/goldfirere/singletons/pull/337#discussion_r195427445) from singling the definition of `transpose` as found in `base`, since it uses partial pattern-matching in a list comprehension (which simply desugars to `do`-notation).

**Comment by goldfirere** [comment]
Oof. That's terrible.

How does this interact with `MonadFail`?

**Comment by RyanGlScott** [comment]
Good question. I suppose we have a choice as to whether we use `Monad.fail` or `MonadFail.fail` in the desugaring. Further complicating matters are that:

1. `MonadFail.fail` is only available in GHC 8.0 or later.
2. GHC currently desugars to `Monad.fail` by default and to `MonadFail.fail` if `-XMonadFailDesugaring` is enabled. We could choose to do the same in `th-desugar`, but unfortunately, the ability to check for the presence of language extensions with Template Haskell was only added in GHC 8.2.

In short, I'm not quite sure which option we should pick, or on which versions of GHC. Thoughts?

**Comment by goldfirere** [comment]
Thanks for that summary. How about this:

Up to and including GHC 8.0: Desugar to `Monad.fail`.
GHC 8.2 and above: consult the extension and the desugar appropriately.

Does that work?

**Comment by RyanGlScott** [comment]
Yes, that sounds reasonable to me. (Upon further inspection, it turns out that `qIsExtEnabled` was added in GHC 8.0, not 8.2, so we only need a single 8.0 CPP boundary.)
